pkg/metrics/metrics.go

```diff
@@ -0,0 +1,135 @@
+
+	// Labels
+	kappNameLabel      = "appName"
+	kappNamespaceLabel = "namespace"
+)
+
+var (
+	sm = serverMetrics{
+		metrics: map[string]prometheus.Collector{
+			reconcileAttemptTotal: prometheus.NewCounterVec(
+				prometheus.CounterOpts{
+					Namespace: metricNamespace,
+					Name:      reconcileAttemptTotal,
+					Help:      "Total number of attempted reconciles",
+				},
+				[]string{kappNameLabel, kappNamespaceLabel},
+			),
+			reconcileSuccessTotal: prometheus.NewCounterVec(
+				prometheus.CounterOpts{
+					Namespace: metricNamespace,
+					Name:      reconcileSuccessTotal,
+					Help:      "Total number of success reconciles",
+				},
+				[]string{kappNameLabel, kappNamespaceLabel},
+			),
+			reconcileFailureTotal: prometheus.NewCounterVec(
+				prometheus.CounterOpts{
+					Namespace: metricNamespace,
+					Name:      reconcileFailureTotal,
+					Help:      "Total number of failed reconciles",
+				},
+				[]string{kappNameLabel, kappNamespaceLabel},
+			),
+			reconcileDeleteAttemptTotal: prometheus.NewCounterVec(
+				prometheus.CounterOpts{
+					Namespace: metricNamespace,
+					Name:      reconcileDeleteAttemptTotal,
+					Help:      "Total number of attempted reconcile deletion",
+				},
+				[]string{kappNameLabel, kappNamespaceLabel},
+			),
+			reconcileDeleteFailedTotal: prometheus.NewCounterVec(
+				prometheus.CounterOpts{
+					Namespace: metricNamespace,
+					Name:      reconcileDeleteFailedTotal,
+					Help:      "Total number of failed reconcile deletion",
+				},
+				[]string{kappNameLabel, kappNamespaceLabel},
+			),
+		},
+	}
+)
+
+// RegisterAllMetrics registers all prometheus metrics.
+func init() {
+	for _, pm := range sm.metrics {
+		metrics.Registry.MustRegister(pm)
+	}
+}
+
+// InitMetrics initializes counter metrics
+func InitMetrics(appName string, namespace string) {
+	for key := range sm.metrics {
+		if c, ok := sm.metrics[key].(*prometheus.CounterVec); ok {
+			c.WithLabelValues(appName, namespace).Add(0)
+		}
+	}
+}
+
+// DeleteMetrics deletes counter metrics
+func DeleteMetrics(appName string, namespace string) {
+	for key := range sm.metrics {
+		if c, ok := sm.metrics[key].(*prometheus.CounterVec); ok {
+			c.DeleteLabelValues(appName, namespace)
+		}
+	}
+}
+
+// RegisterReconcileAttempt increments reconcileAttemptTotal
+func RegisterReconcileAttempt(appName string, namespace string) {
+	if c, ok := sm.metrics[reconcileAttemptTotal].(*prometheus.CounterVec); ok {
+		c.WithLabelValues(appName, namespace).Inc()
+	}
+}
+
+// RegisterReconcileSuccess increments reconcileSuccessTotal
+func RegisterReconcileSuccess(appName string, namespace string) {
+	if c, ok := sm.metrics[reconcileSuccessTotal].(*prometheus.CounterVec); ok {
+		c.WithLabelValues(appName, namespace).Inc()
+	}
+}
+
+// RegisterReconcileFailure increments reconcileFailureTotal
+func RegisterReconcileFailure(appName string, namespace string) {
+	if c, ok := sm.metrics[reconcileFailureTotal].(*prometheus.CounterVec); ok {
```

Comment:
i was looking at antrea metrics [0], and it seems that instead of casting it's better to have that variable be typed.

[0] https://github.com/antrea-io/antrea/blob/main/pkg/agent/proxy/metrics/metrics.go#L29

Reply:
I may not understand this, can you elaborate a bit more ?
BTW, the current structure is inspired a lot from velero metrics.
https://github.com/vmware-tanzu/velero/blob/main/pkg/metrics/metrics.go